### DISCUSSION TITLE
User Brush Editing

### DISCUSSION THREAD

**Discussion by Gemcrafter** [icosa-gallery/open-brush#176] [New and Experimental Features]
Discussion and troubleshooting of custom brushes, made in Unity and/or with the custom brush tool

https://docs.openbrush.app/alternate-and-experimental-builds/experimental-builds/brush-editing

Current issues:
1.  Artifacts appear when created
2.  Tool doesn't appear to identify custom texture files
3.  Issues in working from some custom unity brushes (troubleshooting errors below)

**Comment by Gemcrafter** [comment]
@andybak
andybak

There are four error types appearing when using unity install of integration-branch
And attempting to use the brush tool on a unity created custom brush
![image](https://user-images.githubusercontent.com/7367792/135453081-71d47ab9-6e68-458a-b38f-935218eb6e6c.png)

**Comment by Gemcrafter** [comment]
Errors moved here: https://gist.github.com/andybak/1ad20e5b60dc8dfa26391c55c3cd0a77

**Comment by Gemcrafter** [comment]
First error suggests that read/write needs to be enabled for all imported textures.  Doing that manually now to see if that resolves the secondary issues.   
Probably should be a scripted way to do this

**Reply by andybak** [reply]
Ah yes! I had to do this for all the built-in textures.

It's pretty fast to do via the Unity UI. Not worth the effort of scripting it as it only effects people using it via the Editor.

**Reply by Gemcrafter** [reply]
Isn't fast when you have 500 custom brushes... took 3 album plays... not terribly slow, unless they will reset when I move the brushes
So anyhow changed them manually (unless I missed some).  
Brushes show the modified style in the preview and when making the brush stroke, until the trigger is released to finalize the stroke
The deployed stroke is unchanged.
Do not appear to be saving.  (not seeing in panels or in the brushes folder.... either location)
 Still getting errors as follows:

**Comment by Gemcrafter** [comment]
Further testing.  Reloaded everything again, and the issue appears to be resolved mostly.  
Worked on several custom brushes, including hulls.  Brushes now appear to be saving mostly correct

Did have an issue with an empty preview slot (possibly non-updated for read/write icon, possibly due to action 6)

Need to narrow down the error trigger(s), guessing it's one of the following (since I looked at all these earlier)
1.  A brush with a panel that extended beyond the gui
2.  Attempt to modify a broken brush or brush with unusual parameters (digital or sparks)
3.  Inadvertently modification of an in-game modified custom brush
4.  Attempt to modify a saved file prior to read/write error fix
5.  Attempt to modify a variant of an 'in game' variant brush
6.  Deletion of a created brush inside the brushes folder (had several duplicates earlier).  

Should be able to narrow that down further over the next session or two

**Comment by Gemcrafter** [comment]
Note:  Consider brush previews, generated or using cropped camera shots.  Visual icon does introduce a style difference in the menu, but is easier for an artist to identify brushes.  Worth looking into.

**Comment by Gemcrafter** [comment]
Issue:
Modifying an in-game modified and saved brush does not appear to allow permanent save
While in game, it does appear to take on the characteristics added in the panel, 
however reverts on reload.

Preferred behavior, save as a new variant brush.  (do not want to overwrite the other in-game modified brush)
Potential short term workaround, remodify the original base brush.
Drawback is that any textures and settings used in the modified brush will have to be remembered

Similar brushes are useful when doing complex texturing such as ground surfaces.
*Tested brush was a standard 2d brush that had been modified

Ideally the newly modified brush does not have any dependencies made to the in-game modified brush it is derived from
As at some point will want the ability to delete created brushes as these will likely stack up pretty quickly

**Comment by andybak** [comment]
Error logs moved here: https://gist.github.com/andybak/97f03f9e53cf6d54aef870b27697bc0f

as they were making it hard to read the discussion

**Comment by Gemcrafter** [comment]
Brush save errors appear to be tied to preview icons not having read/write enabled. 
Scrollbars on extended brush GUIs or extended panel needs to be done  (unable to select outside of panel)

New Error ""ArgumentException: C:\Users\shape\Documents\OB5\Brushes is a directory"" .... see error logs
unclear if this is related to custom install, haven't tested with a default install location

Tool massively increases iteration speed.  
Sets with high numbers of icons are likely.   
Texture selection either needs to be subdivided or have a better advancing mechanism (scroll bar or multipage increments option.
Using arrows requires numerous trigger clicks reducing speed and causing index finger/wrist fatigue

Really impressed with what this tool can do in terms of brush design

**Comment by Gemcrafter** [comment]
Pretty long series of tests so will summarize here.  Specific errors detailed in the error section 
| TEST 1
6.  A folder in documents/open brush to dump textures into that can be read from the brush making tool in game
Tested this.  Images in the (c:/documents/brushes) folder are not located in either the default or unity version
(tested top level and as subfolder)
:: occurs to me that empty slots (and consequently save issues) would not happen if a thumbnail is not found
with default brushes this wouldn't be an issue as they use the core thumbnails for icons
[Note: this test was done before read/write enabling.  Will retest with those active] 
AncientWorlds
 — 
Today at 5:41 PM
| Test 2
1.  Made a new unity brush (no other customs yet)
[forgot to unset read only on the texture]
2. brush works on load: error
UnityException: Texture 'colortex2' is not readable, the texture memory can not be accessed from scripts. You can make the texture readable in the Texture Import Settings.
Despite error a new 'in game' brush was generated and was modifiable
two instances of in game brushes were created (possible clicked twice, but don't think so)
3. Modified the bump map.  Appears to work [need to test persistence on reload]
Additional erorr "invalidOpearation Exception nothing to do" 
Error details (2 errors) https://gist.github.com/andybak/97f03f9e53cf6d54aef870b27697bc0f 
AncientWorlds
 — 
Today at 5:51 PM
| Test3 
(Reload program without fixing read/write) 
Original unity created unity brush works
Neither modified brush changed
(multiple instances) NullReferenceException: Object reference not set to an instance of an object
Error detail added (third error at end of thread, no boldface type) https://gist.github.com/andybak/97f03f9e53cf6d54aef870b27697bc0f
AncientWorlds
 — 
Today at 6:14 PM
~
|  Test4
(Fix texturee read/write status and reload)

1. on load: no effect
2. on select original unity brush: no effect
3. on select 'modify brush' : no effect
4. on deselect 'modify brush': no effect
5. on select modified brush: error occurs
NullReferenceException: Object reference not set to an instance of an object
error 4

Summary of error 4
occurs when modified brush made without read/write is accessed
after read/write has been toggled on restart 
AncientWorlds
 — 
Today at 6:32 PM
| Test 5
(make a new modified brush of the unity built clone, inside the game engine)
(this modified brush is being made post read/write fix in textures)

1. load game: no error
2. make brush stroke with custom brush: no error
3. Select 'modify brush': Error (5) [error documented in error thread]
     Innvalid Operation Exception: Nothing to do 
4. change bumpmap: no error
5: save: no error (new brush created successfully)
6. Switch to newly created brush: no error
7.  Select bumpmap.  :no error
8. Save:  Error (6) 
    UnityException: Texture 'colortex2' is not readable, the texture memory can not be accessed from scripts. You can make the texture readable in the Texture Import Settings.
    //colortex2 is my custom maintexture
   //   Used the leaf bumpmap as replacement texture 

Replicated:
Verified that read write enable was active
 — 
|  Test 6
:: went to test read/writes in documents folder
observed I had copied colortex2.png into there and not removed (probably not read/write enabled)
deleted it. 

Attempted to replicate test 5, error did not occur.  [SEE TEST 9, was able to replicate]
used different replacement texture, will reverify 
| Test 7 
1. opened program
2. selected the most recent in-game custom brush
3. did not toggle the brush edit tool in panel
4. made brush stroke (so it used the default 'light' brush) :  Error (7) multiple instances
    NullReferenceException: Object reference not set to an instance of an object
    
| Test 8: 
verified leaf texture was read/write enabled.  It is.  It is also a normal map 

Test 9:
attempt to replicate test 5
successfully replicated error result 
 — 

| Test 10:
attempt to replicate previous test post reload 
// redo test, lost tracking.  All VR and unity exited 
// nothing to do error. // colortex2 is not readable 
::  Entire sequence described above placed here
https://github.com/icosa-gallery/open-brush/discussions/176
So there's the details, step by step on most of the tests.
possibly multiple issues.  Is there a specific test variant you'd like to see?
:: As  a control I'm going to make a similar brush (identical except maps) and ensure that all read/writes are enabled prior to launching program

 — 

|Test 11:
1. Rebuilt brush and material using same paramaters different images
2. prior to creating material all read/writes enabled and verified
3.  Launch program
4.  Select custom brush : no error
5. Select modify brush tool: Nothing to do error [Error 8)
6.  (proceed)  test brush stroke: no error
7.  change texture map:  no error
8.  Save:  No error
9.  Switch to modified brush : no error
10.  Change bump_map : no error  [using a default available map]
11.  Save | [Error 9]
    UnityException: Texture 'main' is not readable, the texture memory can not be accessed from scripts. You can make the texture readable in the Texture Import Settings.
